source/common/formatter/http_specific_formatter.cc

```diff
@@ -294,6 +294,83 @@ QueryParameterFormatter::formatValueWithContext(const HttpFormatterContext& cont
   return ValueUtil::optionalStringValue(formatWithContext(context, stream_info));
 }
 
+absl::optional<std::string> PathFormatter::formatWithContext(const HttpFormatterContext& context,
+                                                             const StreamInfo::StreamInfo&) const {
+
+  absl::string_view path_view;
+  const Http::RequestHeaderMap& headers = context.requestHeaders();
+  switch (option_) {
+  case OriginalPathOrPath:
+    path_view = headers.getEnvoyOriginalPathValue();
+    if (path_view.empty()) {
+      path_view = headers.getPathValue();
+    }
+    break;
+  case PathOnly:
+    path_view = headers.getPathValue();
+    break;
+  case OriginalPathOnly:
+    path_view = headers.getEnvoyOriginalPathValue();
+    break;
+  }
+
+  if (path_view.empty()) {
+    return absl::nullopt;
+  }
+
+  // Strip query parameters if needed.
+  if (!with_query_) {
+    auto query_offset = path_view.find('?');
+    if (query_offset != absl::string_view::npos) {
+      path_view = path_view.substr(0, query_offset);
+    }
+  }
+
+  // Truncate the path if needed.
+  if (max_length_.has_value()) {
+    path_view = SubstitutionFormatUtils::truncateStringView(path_view, max_length_);
+  }
+
+  return std::string(path_view);
+}
+
+ProtobufWkt::Value
+PathFormatter::formatValueWithContext(const HttpFormatterContext& context,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  return ValueUtil::optionalStringValue(formatWithContext(context, stream_info));
+}
+
+absl::StatusOr<FormatterProviderPtr> PathFormatter::create(absl::string_view with_query,
+                                                           absl::string_view option,
+                                                           absl::optional<size_t> max_length) {
+  bool with_query_bool = true;
+  PathFormatterOption option_enum = OriginalPathOrPath;
+
+  if (with_query == "WQ") {
+    with_query_bool = true;
+  } else if (with_query == "NQ") {
+    with_query_bool = false;
+  } else if (with_query.empty()) {
+    with_query_bool = true;
+  } else {
+    return absl::InvalidArgumentError(fmt::format("Invalid PATH option: {}", with_query));
```

Comment:
In the error string, say what valid values are (or maybe link to the docs page?)